refactor(iroh)!: remove Endpoint::add_node_addr

dignifiedquire opened the pull request:
## Description

Simplifies the public API. If you need to provide static information about node addresses, please use `StaticProvider` from now on.

Closes https://github.com/n0-computer/iroh/issues/3329

## Breaking Changes

- removed
  - `iroh::Endpoint::add_node_addr`
  - `iroh::Endpoint::remote_info`
  - `iroh::Endpoint::remote_info_iter`
  - `iroh::RemoteInfo`
  - `iroh::discovery::DiscoveryItem`
- added
  - `iroh::discovery::mdns::DiscoveryItem`
  - `iroh::Endpoint::latency`

## Notes & open questions

Is it okay to have these all removed now, given the replacements for connection details have not yet landed (will only with multipath)?

rklaehn commented:
I thought this might break big-block-syc, but it seems like in the latest version of big-block-sync I only use Connection::rtt(). So I guess it's fine. I will have to refactor a lot of code to get rid of add_node_addr, e.g. in the dht experiment, but it should be fine.

Not a fan of having any kind of mutable default StaticDiscovery in the endpoint. Just accept the inconvenience instead of creating chaos later when multiple places try to modify it.

https://github.com/n0-computer/big-block-sync/blob/main/src/lib.rs

dignifiedquire commented:
This currently breaks the advertised way to see (passively) discovered MDNS nodes, need to figure this part out.

dignifiedquire commented:
Latency should be fine, `.rtt()` on the connection is preferrable anway

ramfox commented:
> This currently breaks the advertised way to see (passively) discovered MDNS nodes, need to figure this part out.

Yea, I think the only option is to change the example to show you need to build the mdns discovery service, subscribe to the stream, and THEN add the discovery service to the endpoint, now that we can add discovery services after the endpoint has been built.

ramfox commented:
Hmmm, to @Frando 's point, I think as long as the docs specify that the `remote_info` and `remote_info_iter` returns information for remote endpoints that we have called called `connect` on or have accepted connections from, then it's valid to add those methods back in.

ramfox commented:
Cleaned things up so tests are passing:
  - adjusted doc comments and examples
  - removed `discovery_subscriber` method and `DiscoverySubscriber` struct

API thoughts:
1) The current way to `add` a discovery service after the endpoint has been created is:
`ep.discovery().clone().add(new_discovery_service)`
The `clone` is a bit of a footgun, because the error that the complier gives you is about the `ConcurrentDiscovery` being behind a `&`, so it can't be mutable.
Now that `ConcurrentDiscovery` has an `Arc`, should we just return a `ConcurrentDiscovery` rather than a `&ConcurrentDiscovery`, from `ep.discovery()`?

2) Now that `subscribe` is only an `MdnsDiscovery` method, we should likely change the return types from `DiscoveryEvent` to `MdnsEvent`. Also, we now have to including `tokio::pin`-ing the stream, why did we not have to do that before?
  
3) I do think we might be foot-gunning ourselves by removing `remote_info`/`remote_info_iter` too early until we have a valid replacement.

dignifiedquire commented:
> Also, we now have to including `tokio::pin`-ing the stream, why did we not have to do that before?

yeah, a bit annoying, I want to try and fix it

dignifiedquire commented:
> The current way to add a discovery service after the endpoint has been created is:
ep.discovery().clone().add(new_discovery_service)

fixed

> Now that subscribe is only an MdnsDiscovery method, we should likely change the return types from DiscoveryEvent to MdnsEvent

fixed

> Also, we now have to including `tokio::pin`-ing the stream, why did we not have to do that before?

fixed

flub commented:
:tada:
